graders/c/cgrader/cgrader.py

```diff
@@ -0,0 +1,217 @@
+#! /usr/bin/python3
+
+import os, json, subprocess, shlex, threading, re, sys
+
+CODEBASE = '/grade/student'
+DATAFILE = '/grade/data/data.json'
+SB_USER = 'sbuser'
+
+class CGrader:
+
+    def __init__(self):
+        with open(DATAFILE) as file:
+            self.data = json.load(file)
+
+    def run_command(self, command, input=None, sandboxed=False, timeout=None):
+        if isinstance(command, str):
+            command = shlex.split(command)
+        if sandboxed:
+            command = ['su', SB_USER, '-s', '/bin/bash', '-c', shlex.join(['PATH=' + self.path] + command)]
+        try:
+            proc = subprocess.Popen(command,
+                                    stdin=subprocess.PIPE,
+                                    stdout=subprocess.PIPE,
+                                    stderr=subprocess.STDOUT,
+                                    encoding='utf-8')
+        except:
+            return ''
+        out1 = out2 = None
+        tostr = ''
+        try:
+            out1, err = proc.communicate(input=input, timeout=timeout)
+        finally:
+            proc.kill()
+            try:
+                out2, err = proc.communicate(timeout=timeout)
+            except subprocess.TimeoutExpired:
+                tostr = '\n\nTIMEOUT! Typically this means the program took too long,' + \
+                        '\nrequested more inputs than provided, or an infinite loop was found.' + \
+                        '\nIf your program is reading data using scanf inside a loop, this ' + \
+                        '\ncould also mean that scanf does not support the input provided ' + \
+                        '\n(e.g., reading an int if the input is a double).'
+            finally:
+                out = (out1 if out1 else '') + (out2 if out2 else '') + tostr
+                return out
+
+    def test_compile_file(self, c_file, exec_file, main_file=None, points=1, field=None, name='Compilation'):
```

Comment:
It's possible to have autograders return an invalid/non-gradable score for the grading, which doesn't use up retry-attempts on a question. Some people have been using this for cases when the code doesn't compile, so students don't get penalized in this case. I'm not sure whether that's something you want to consider here? I don't have a strong opinion either way, I'm just raising it as a possibility.

If you want to do this, you need to set `"gradable": false` in `results.json`. See https://prairielearn.readthedocs.io/en/latest/externalGrading/#grading-results

Reply:
Great suggestion, and I even set it as default, and caused an exception to stop other tests from running. Students with code that doesn't compile were focusing on the fact that tests failed instead of noticing the errors.